packages/web-components/fast-ssr/src/template-parser/template-parser.ts

```diff
@@ -0,0 +1,351 @@
+     * below, so store in a new var that is just a string type
+     */
+    const templateString = html;
+    const nodeTree = parseFragment(html, { sourceCodeLocationInfo: true });
+
+    if (!("nodeName" in nodeTree)) {
+        // I'm not sure when exactly this is encountered but the type system seems to say it's possible.
+        throw new Error(`Error parsing template:\n${template}`);
+    }
+
+    /**
+     * Tracks the offset location in the source template string where the last
+     * flushing / skip took place.
+     */
+    let lastOffset: number | undefined = 0;
+
+    /**
+     * Collection of op codes
+     */
+    const opCodes: Op[] = [];
+    opCache.set(template, opCodes);
+
+    const { directives } = template;
+
+    /**
+     * Parses an Element node, pushing all op codes for the element into
+     * the collection of ops for the template
+     * @param node - The element node to parse
+     */
+    function parseElementNode(node: DefaultTreeElement): void {
+        // Track whether the opening tag of an element should be augmented.
+        // All constructable custom elements will need to be augmented,
+        // as well as any element with attribute bindings
+        let augmentOpeningTag = false;
+        const { tagName } = node;
+        let ctor: typeof HTMLElement | undefined;
+
+        // Sort attributes by whether they're related to a binding or if they have
+        // static value
+        const attributes: {
+            static: Map<string, string>;
+            dynamic: Attribute[];
+        } = node.attrs.reduce(
+            (prev, current) => {
+                if (isInterpolationMarker(current)) {
+                    prev.dynamic.push(current);
+                } else {
+                    prev.static.set(current.name, current.value);
+                }
+
+                return prev;
+            },
+            { static: new Map(), dynamic: [] as Attribute[] }
+        );
+
+        // Special processing for any custom element
+        if (node.tagName.includes("-")) {
+            ctor = customElements.get(tagName);
+
+            if (ctor !== undefined) {
+                augmentOpeningTag = true;
+                node.isDefinedCustomElement = true;
+                opCodes.push({
+                    type: OpType.customElementOpen,
+                    tagName,
+                    ctor,
+                    staticAttributes: attributes.static,
+                });
+            }
+        }
+
+        // Push attribute binding op codes for any attributes that
+        // are dynamic
+        if (attributes.dynamic.length) {
+            for (const attr of attributes.dynamic) {
+                const location = node.sourceCodeLocation!.attrs[attr.name];
+                flushTo(location.startOffset);
+                const attributeType = getAttributeType(attr);
+                const parsed = Parser.parse(attr.value, directives);
```

Comment:
If you use this API above instead, then you'll have this data already and if you call `aggregate`, it will give you a single directive that should have knowledge of the binding target type (property, text, attr, etc). I'm not sure if that's exposed but I could make that available.

Reply:
Yeah that would be great - I wasn't sure how to get that from the directive but if we can, then perfect.

Reply:
It doesn't look like it's exported. I'll write up a task to add that feature to fast-element and then we can fix up the ssr code to use it.